### ISSUE TITLE
Auto-tag H7 headings using Docraptor

### ISSUE THREAD

**Issue by pcraig3** [HHS/simpler-grants-pdf-builder#448]
Early in the life of the NOFO Builder, we decided that every NOFO has _only one_ H1, which is a common rule of accessibility. Some people allow for more than one H1, but this is our rule and we’re sticking to it. 

The problem we ran into is that many authors use content guides that use H1s for the “Step” headings (“Step 1: Review the Opportunity”), whereas in our PDF those become H2s. The H1 is the title of the NOFO. 

This means that for some NOFOs, **we automatically demote all the headings in the Word doc by 1 level**. So H1s become H2s, H2s become H3s, etc. This works until we get to H6, which is the last heading level formally recognized in the PDF standard. So we created an shimmed H7 heading level. 

In HTML, there is a W3C suggestion for shimming them in using “role” and “aria-level’’ attributes: https://www.w3.org/TR/WCAG20-TECHS/ARIA12.html#ARIA12-ex2

Digging around, I found a couple of results on H7s in PDFs:
* [https://taggedpdf.com/508-pdf-help-center/out-of-heading-levels](https://taggedpdf.com/508-pdf-help-center/out-of-heading-levels/)
* [https://pdfa.org/wp-content/uploads/2019/06/TaggedPDFBestPracticeGuideSyntax.pdf](https://pdfa.org/wp-content/uploads/2019/06/TaggedPDFBestPracticeGuideSyntax.pdf) (quoted below) 
    * "PDF/UA-1 requires the use of and higher heading levels if semantically appropriate. Such structure types are undefined in PDF 1.7. PDF/UA requires a mapping for undefined elements; accordingly, in this case it is recommended that and higher levels of headings be role mapped to <p>"

In both cases, the recommendation is to manually tag the elements "H7" and then role-map them to `<P>` tags. 

We are currently role-mapping `H7`s to `P` tags now that [the latest prince](https://www.princexml.com/releases/16/) is out: https://github.com/HHS/simpler-grants-pdf-builder/blob/10ebc22637b27648197f6666723ae30b5e597e44/nofos/bloom_nofos/static/theme-base.css#L1109

But we weren’t sure how they would be interpreted by screen readers. 

I created 3 PDFs to test this. On page 2 of each we have our heading levels from H2 to H7, with a paragraph of text underneath each one. 

1. [TEST_HEADINGS_P](https://drive.google.com/file/d/1S2TC0ppDOWJQtn8rlCxfYaAwPvDljlFE/view?usp=sharing): no special treatment, the H7 should read as a paragraph 
2. [TEST_HEADINGS_H7_MANUAL](https://drive.google.com/file/d/1_HhEnf0XqPn-WdUtQgYYP2Yt5ZpogEmh/view?usp=sharing): We manually tagged the final heading as "H7" 
3. [TEST_HEADINGS_H7_AUTO](https://drive.google.com/file/d/1XwFqqk6a6qe5s__hW6FL7mvVIKElhkyn/view?usp=sharing): We programmatically tagged the final heading as "H7" 

I tested them all on VoiceOver on MacOS Preview and I can't actually tell if there is any difference between them. The other heading levels are announced, the H7 tag is never announced.

We also crowdsourced this by [posting it on Reddit](https://www.reddit.com/r/accessibility/comments/1mq8e0h/need_help_from_someone_who_uses_screenreaders_is/) and received the feedback that it is the same in other screen readers:

> It doesn’t work on any of the PDFs. I tested it on three different screen readers.

So our takeaways are:

- Automatically tagging the PDF is an appropriate substitute for manual tagging
- Regardless, all screen readers just treat them like H7s
- There is some chance that a future PDF standard (2.0?) will recognize custom elements like this, but who knows when that will ever happen.

So that’s it! 

We are not automatically tagging H7s in PDFs, but we also don’t think that screen readers pick that up.

After like a year of kicking this around, the Issue is closed.

**Comment by pcraig3** [comment]
Closed since #456 went in.
